pkg/logs/tag/provider.go

```diff
@@ -6,78 +6,50 @@
 package tag
 
 import (
-	"reflect"
 	"sync"
 	"time"
 
+	"github.com/DataDog/datadog-agent/pkg/logs/config"
 	"github.com/DataDog/datadog-agent/pkg/tagger"
 	"github.com/DataDog/datadog-agent/pkg/tagger/collectors"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
-const refreshPeriod = 10 * time.Second
-
 // Provider returns a list of up-to-date tags for a given entity.
 type Provider interface {
 	GetTags() []string
-	Start()
-	Stop()
 }
 
 // NoopProvider does nothing
 var NoopProvider Provider = &noopProvider{}
 
-// provider caches a list of up-to-date tags for a given entity polling periodically the tagger.
+// provider provides a list of up-to-date tags for a given entity by calling the tagger.
 type provider struct {
-	entityID string
-	tags     []string
-	done     chan struct{}
-	mu       sync.Mutex
+	entityID             string
+	taggerWarmupDuration time.Duration
+	sync.Once
 }
 
 // NewProvider returns a new Provider.
 func NewProvider(entityID string) Provider {
 	return &provider{
-		entityID: entityID,
-		tags:     []string{},
-		done:     make(chan struct{}),
+		entityID:             entityID,
+		taggerWarmupDuration: config.TaggerWarmupDuration(),
 	}
 }
 
 // GetTags returns the list of up-to-date tags.
 func (p *provider) GetTags() []string {
-	p.mu.Lock()
-	defer p.mu.Unlock()
-	return p.tags
-}
-
-// Start starts the polling of new tags on another go routine.
-func (p *provider) Start() {
-	go func() {
-		p.updateTags()
-		ticker := time.NewTicker(refreshPeriod)
-		defer ticker.Stop()
-		for {
-			select {
-			case <-ticker.C:
-				p.updateTags()
-			case <-p.done:
-				return
-			}
-		}
-	}()
-}
-
-// Stop stops the polling of new tags.
-func (p *provider) Stop() {
-	p.done <- struct{}{}
-}
-
-// updateTags updates the list of tags using tagger.
-func (p *provider) updateTags() {
-	p.mu.Lock()
-	defer p.mu.Unlock()
+	p.Do(func() {
+		// Warmup duration
+		// Make sure the tagger collects all the service tags
+		// TODO: remove this once AD and Tagger use the same PodWatcher instance
+		<-time.After(p.taggerWarmupDuration)
```

Comment:
How would this behave with concurrency?

Let's say we call this twice in separate routines, the first call would get delayed but wouldn't the second one go through without the warmup being over, or am I missing something?

Reply:
> Do calls the function f if and only if Do is being called for the first time for this instance of Once
(https://golang.org/pkg/sync/#Once) 

In our case, we create one config provider instance per AD service (entity).

Reply:
So we guarantee that every first `GetTags` call for every Tag Provider will warmup

Reply:
and for each AD service I guess the call is blocking other log lines from being processed right? (no concurrency inside a service)

Reply:
yes [forwardMessages()](https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/input/file/tailer.go#L116) runs in a single go routine (per service) and [doesn't](https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/input/file/tailer.go#L226-L254) create any other go routine 

Same for the [docker tailer](https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/input/docker/tailer.go)

@ajacquemot please correct me if I'm wrong

Reply:
this is all correct, we could add a comment which states that this method is not thread safe and should not be called concurrently